### PR TITLE
Add support for special methods

### PR DESCRIPTION
**Pull request by Manishearth**
Progress on https://github.com/rust-diplomat/diplomat/issues/447

Does not yet use them in Dart/C++. Also doesn't touch iterators (#251) since that requires the type annotation to be a thing first.

### CODE REVIEW

File: core/src/hir/attrs.rs [hunk 1]
```diff
@@ -2,45 +2,89 @@
...
+pub enum SpecialMethod {
+    /// A constructor.
+    ///
+    /// Must return Self (or Result<Self> for backends with `fallible_constructors` enabled )
+    Constructor,
+    /// A named constructor, with optional name. If the name isn't specified, it will be derived
+    /// from the method name
+    ///
+    /// Must return Self (or Result<Self> for backends with `fallible_constructors` enabled )
+    NamedConstructor(Option<String>),
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply core/src/hir/attrs.rs:55]: There is one already

File: example/src/decimal.rs [hunk 2]
```diff
@@ -47,6 +48,7 @@ pub mod ffi {
     impl ICU4XFixedDecimalFormatter {
         /// Creates a new [`ICU4XFixedDecimalFormatter`] from locale data.
         #[diplomat::rust_link(icu::decimal::FixedDecimalFormatter::try_new, FnInStruct)]
+        // TODO constructors: this should ideally be a constructor too
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply example/src/decimal.rs:51]: it returns null

File: core/src/hir/attrs.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (robertbastian) [reply core/src/hir/attrs.rs:55]: It's not documented here.

File: example/src/decimal.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (robertbastian) [reply example/src/decimal.rs:51]: ah void error

File: feature_tests/dart/lib/src/ResultOpaque.g.dart [hunk 3]
```diff
@@ -35,7 +35,7 @@ final class ResultOpaque implements ffi.Finalizable {
   /// 
   ///
   /// Throws [ErrorEnum] on failure.
-  factory ResultOpaque.failingFoo() {
+  factory ResultOpaque.newFailingFoo() {
```
Comment (robertbastian) [inline feature_tests/dart/lib/src/ResultOpaque.g.dart:38]: hmm where does this `new` come from? shouldn't exist in Dart

File: example/src/decimal.rs [hunk 4]
```diff
@@ -36,6 +36,7 @@ pub mod ffi {
     }
 
     impl ICU4XFixedDecimalFormatterOptions {
+        #[diplomat::attr(supports = constructors, constructor)]
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply example/src/decimal.rs:39]: doable, should be a followup.

[the remark this answers is not in the document]
Comment (Manishearth) [reply example/src/decimal.rs:39]: https://github.com/rust-diplomat/diplomat/issues/453

**Pull request merged**